### PR TITLE
Port vainfo and tests for libva-win32

### PR DESCRIPTION
**Pull request by sivileri**
This MR ports vainfo and the gtest suite for Windows using libva-win32 

Requires: https://github.com/intel/libva/pull/621

### CODE REVIEW

**Comment by evelikov** [comment]
As a draft this is great. For the final PR please consider:
  - keeping the getopt a separate commit with references (source and version/git commit) in the commit message
  - adding support for non-default (NULL) display
  - run through style-unify, if you haven't already
  - porting at least some of the following - decode, encode, videoprocess and tests - hopefully they should be a matter of a trivial `compat_win32.h` header

The final pieces don't make sense to port IMHO:
 - vendor/intel - there is no intel driver for win32 (yet at least)
 - putsurface - the API is not implemented in va-win32, rightfully so IMHO

**Comment by sivileri** [comment]
@evelikov thanks for taking a look, please see inline.
> As a draft this is great. For the final PR please consider:
> 
> * keeping the getopt a separate commit with references (source and version/git commit) in the commit message

[sivileri] Will do.

> * run through style-unify, if you haven't already

[sivileri] Got it. I ran it through the files I changed, no changes needed. 

> * adding support for non-default (NULL) display
> * porting at least some of the following - decode, encode, videoprocess and tests - hopefully they should be a matter of a trivial `compat_win32.h` header

[sivileri] I tried doing this earlier today, I could use a similar win32_compat like the one used for the libva repo for some of the it like the time.h functions and variadic macros, but in here the changes need to be a bit more invasive and will change significant parts of the existing code, the biggest reason being the explicit initialization with `<fieldname> : <value>` that's used for structs in the tests and doesn't build with MSVC, some tests using pthreads among others and being pure C limits the usage of modern features like COM RAII for D3D code, etc.

To avoid doing such a disruptive change on existing code, and considering there are open source samples for VA on Windows [here](https://github.com/sivileri/DirectX-Graphics-Samples/tree/video_sample/Samples/Desktop/D3D12HelloWorld/src/HelloVAEncode ) that do adapter enumeration and selection, encode, video processing, importing/exporting of surfaces and interop with D3D12 to present surfaces to screen, maybe we can add those to this repo and/or link them here to the project where those live? I'll work on porting a decode example in that sample app for completeness.
> 
> The final pieces don't make sense to port IMHO:
> 
> * vendor/intel - there is no intel driver for win32 (yet at least)
> * putsurface - the API is not implemented in va-win32, rightfully so IMHO

[sivileri] Sounds good.

**Comment by evelikov** [comment]
> the biggest reason being the explicit initialization with <fieldname> : <value> that's used for structs in the tests and doesn't build with MSVC

Off the top of my head, tests/gtest should just work with MSVC, do shout if it does not. Actual tests like `test_va_api_get_create_config.cpp` should really be fixed to use standard C++ by either dropping the `<fieldname> :` or using `C++20` designated initialises (the ones we got with C99).

> some tests using pthreads among others and being pure C limits the usage of some features like COM RAII for D3D code, etc.

Not familiar with the details off the top of my head. If you have references I might come-up with some half-crazy idea :-)

> ... maybe we can add those to this repo and/or link them here to the project where those live? I'll work on porting a decode example in that sample app for completeness.

No idea if copying these into the repo will try. Although having at least one actual example ported (apart from vainfo of course) - decode or others, would be amazing IMHO.

File: common/va_display_win32.c
Comment (dvrogozh) [inline common/va_display_win32.c:1]: Add license header, please.

File: getopt/getopt.h
Comment (dvrogozh) [inline getopt/getopt.h:1]: See: https://github.com/Mesa3D/mesa/commit/f9bb5323eb96f47cfb4ab5f93165323df0a1fd61

**Comment by evelikov** [comment]
Glancing at the in-repo tools - they have been bit-rotting since 2016 when they were split out of libva itself. In libva we've had a dozen APIs and even more enum additions, while this repo has predominantly seen static analyser fixes. There is one notable exception - porting the tests to GTest \o/

Now I see why you saw more value in writing fresh HelloWorld demos instead of digging into the graveyard :+1: 

Personally I would strongly recommend porting the tests and not worrying about the rest. They should be trivial to fix, plus the Mesa CI should be using them already (or is about to soon).

**Comment by sivileri** [comment]
> Glancing at the in-repo tools - they have been bit-rotting since 2016 when they were split out of libva itself. In libva we've had a dozen APIs and even more enum additions, while this repo has predominantly seen static analyser fixes. There is one notable exception - porting the tests to GTest \o/
> 
> Now I see why you saw more value in writing fresh HelloWorld demos instead of digging into the graveyard 👍
> 
> Personally I would strongly recommend porting the tests and not worrying about the rest. They should be trivial to fix, plus the Mesa CI should be using them already (or is about to soon).

I went ahead and ported the tests too now, I was able to build and run them.

**Comment by evelikov** [comment]
Incoming some nitpicks and a question for the mid-long run. Overall I'm in favour of the PR and don't see anything being off.

File: meson.build
Comment (evelikov) [inline meson.build:80]: Mid/long term question:

Ideally we'd want va-win32 detection and building to work with mingw on top of Linux. Both here and in libva itself.

File: meson.build
Comment (evelikov) [inline meson.build]: Nits:
 - typo s/win32s/win32/
 - getopt cannot be "not_found()", so I would flatten the foreach loop below

File: test/test_utils.h
Comment (evelikov) [inline test/test_utils.h:42]: Nit: I would keep these and/or -> && and || as separate commit. same for the gnu initializers further down

File: test/test_va_api_fixture.cpp
Comment (evelikov) [inline test/test_va_api_fixture.cpp:57]: We're changing from `char *` to a fixed size string. Might be wroth a separate commit with some details why

Comment (evelikov) [reply test/test_va_api_fixture.cpp:57]: Final nit - if it were me, I would prefix the commits with the section they touch.

For example:
vainfo: add va-win32 support
tests: remove non C++ standard constructs
tests: enable building on Windows

File: meson.build
Comment (sivileri) [reply meson.build]: Sounds good. Enabling usage of win32 should only depend on libva-win32 being present, disregarding the platform

File: test/test_va_api_fixture.cpp
Comment (sivileri) [reply test/test_va_api_fixture.cpp:60]: I broke down the commits more for clarity :)

Comment (evelikov) [reply test/test_va_api_fixture.cpp:60]: > It is not safe to modify the value of the environment variable using the returned pointer

We don't do that, do we?

Comment (sivileri) [reply test/test_va_api_fixture.cpp:60]: Yes, in `VAAPIFixture::~VAAPIFixture` it restores the value using `setenv("LIBVA_DRIVER_NAME", m_restoreDriverName, 1);`.
Passing `m_restoreDriverName` to `setenv` with it being the exact pointer returned from `getenv` produces sets the environment variable to garbage (I can confirm running the tests when vainfo prints the driver name in screen)

Comment (evelikov) [reply test/test_va_api_fixture.cpp:60]: According to `man setenv`
```
int setenv(const char *name, const char *value, int overwrite);
```

In other words - the string is not modified and the current comment is misleading at best.

Perhaps Windows `setenv()` modifies it, or it has different lifetime there? Don't think the "why" matters too much - just document that you're seeing garbage on Windows and let's merge this already - hint maintainers, hint :wink: :smiling_imp:

File: test/test_va_api_fixture.cpp
Comment (dvrogozh) [inline test/test_va_api_fixture.cpp:63]: That's C++ code, how about:
```
std::string m_restoreDriverName; // in constructor, evntually

char* libva_driver = getenv("LIBVA_DRIVER_NAME");
if (libva_driver) m_restoreDriverName = libva_driver;
...
setenv("LIBVA_DRIVER_NAME", m_restoreDriverName.c_str(), 1);

```

File: test/test_va_api_fixture.cpp
Comment (dvrogozh) [inline test/test_va_api_fixture.cpp:360]: why this change was made?

File: test/test_va_api_fixture.cpp
Comment (dvrogozh) [inline test/test_va_api_fixture.cpp:531]: 1. Why you did not modify `TearDownTestSuite`? It seems you are missing `vaTerminate` for windows.
2. I wish you would minimize complex macro constructions. For example, you don't exclude `s_drmHandle` and you can use that even on Windows:
```
void VAAPIFixtureSharedDisplay::SetUpTestSuite()
{
    if (s_drmHandle < 0) {
#if defined(_WIN32)
        s_vaDisplay = getWin32Display(NULL);
#else
        s_vaDisplay = getDrmDisplay(s_drmHandle);
#endif
        int majorVersion, minorVersion;
        s_initStatus = vaInitialize(s_vaDisplay, &majorVersion, &minorVersion);
    }
}

void VAAPIFixtureSharedDisplay::TearDownTestSuite()
{
    if (s_vaDisplay) {
        if (s_initStatus == VA_STATUS_SUCCESS) {
            vaTerminate(s_vaDisplay);
            s_vaDisplay = nullptr;
        } else {
            s_initStatus = VA_STATUS_SUCCESS;
        }
    }
    if (s_drmHandle >= 0) {
        close(s_drmHandle);
        s_drmHandle = -1;
    }
}
```

**Review by dvrogozh — CHANGES_REQUESTED** [review]

**Comment by dvrogozh** [comment]
> let's merge this already - hint maintainers, hint 

We are waiting for @XinfengZhang to try libva and libva-utils changes. This won't happen earlier than next week due to holidays in China.

File: meson.build
Comment (sivileri) [reply meson.build:80]: false will disable it explicitly, true will try to force enable it and fail if dependencies not found. 'auto' (which is also what you get if you do not specify the arguments in the meson commandline) will try to enable it if the dependencies are found. In other words, if you run "meson build/" on windows it will enable win32 if it finds libva-win32

File: test/test_va_api_fixture.cpp
Comment (sivileri) [reply test/test_va_api_fixture.cpp:360]: There's a `#define max(a,b)            (((a) > (b)) ? (a) : (b))` definition in `minwindef.h` which gets pulled when building on Windows and conflicts with that syntax, producing a build fail otherwise. I will add a comment to it.

Comment (dvrogozh) [reply test/test_va_api_fixture.cpp:357]: I know about this definition, but I don't understand how your change helps to avoid it? Basically you still have `max` in the code and it should get substituted producing something which should not compile, i.e.: 
```
auto maxFn = std::numeric_limits<Resolution::DataType>::(((a) > (b)) ? (a) : (b));
```

In other projects we were resolving this problem by adding `NOMINMAX` which from my perspective is a better way.

File: meson.build
Comment (dvrogozh) [inline meson.build:83]: I always thought that `/` options syntax is something very much windows specific. Will '/std:c++20' actually work with mingw?

Also, is there any standard way to specify required c++ standard in meson?

Comment (sivileri) [reply meson.build]: It does build on both MSVC and MINGW. The way to specify std in meson is in libva-utils/meson.build in the project() tag, adding 'cpp_std=c++XX'. The original project is targetting C++11 in linux (test/meson.build has -std=c++11) so I just went ahead and replaced the only small C++20 usage with C++11 compatible code and was able to remove this altogether.

**Comment by sivileri** [comment]
@dvrogozh @XinfengZhang now that https://github.com/intel/libva/pull/621 is merged, could you please enable the workflows execution? I'll work on making them work with the libva-win32 dependency from libva/master

image

**Comment by sivileri** [comment]
> @dvrogozh @XinfengZhang now that [intel/libva#621](https://github.com/intel/libva/pull/621) is merged, could you please enable the workflows execution? I'll work on making them work with the libva-win32 dependency from libva/master
> 
> image

There we go: https://github.com/sivileri/libva-utils/actions/runs/3244558566/jobs/5320888674 :) Please let me know if I can do anything else before we can merge this MR too.

image

**Comment by dvrogozh** [comment]
> now that https://github.com/intel/libva/pull/621 is merged, could you please enable the workflows execution?

Approved. But mind that this is a manual action which needs to be done each time till you won't have at least 1 commit merged:). So, if you will push an update it will need a new approval.

**Review by dvrogozh — CHANGES_REQUESTED** [review]

**Comment by dvrogozh** [comment]
freebsd build fails (you need to open "test" stage to see actual error):
```
ld: error: undefined symbol: va_display_hooks_win32
```

**Review by dvrogozh — APPROVED** [review]
looks good

What did you do to fix freebsd build? Can you point to the change, please?

**Comment by evelikov** [comment]
> What did you do to fix freebsd build? Can you point to the change, please?

The "Compare" button shows them - aka `DHAVE_VA_WIN32=1` was incorrectly set.

**Comment by sivileri** [comment]
> looks good
> 
> What did you do to fix freebsd build? Can you point to the change, please?

Thanks! DHAVE_VA_WIN32=1 was being always set disregarding of the dependency check. Here's the fix from the latest push:
https://github.com/intel/libva-utils/compare/b9d961ecd78b5c742aa834abbb4e3c775cb70e9d..88cd38ba4834c6da42d7084444fc6adbab9affdc
image

**Comment by dvrogozh** [comment]
Thank you for pointing out. I am fine with the change. @XinfengZhang : can you, please, take a look and merge?

File: meson.build
Comment (XinfengZhang) [inline meson.build:100]: is it possible to make sample codes work on windows?

File: test/test_utils.h
Comment (XinfengZhang) [inline test/test_utils.h:42]: these changes just because VS does not support iso646?

**Review by XinfengZhang — APPROVED** [review]

File: test/test_utils.h
Comment (sivileri) [reply test/test_utils.h:42]: Yes, the ciso646.h header will be deprecated too so I thought it was easier to use the non-iso646 symbols

File: meson.build
Comment (sivileri) [reply meson.build:100]: For most of them we don't have driver support yet to test it (ie. mpeg2, jpeg decode, vp8/vp9 encode and most VPP effects tested). I think the ones we could port and test today are: encode H264/HEVC and VPP scaling and color conversion (only NV12, P010 and RGBX supported in driver at the moment). 

From those we could port and test at the moment, cross-platform porting is needed, the most notable port needed would be `pthreads` for the encode samples. 

As we're also in the process of writing new samples for Windows that cover those same Decode/Encode/VPP scenarios implemented by the driver today, also showing D3D12/Libva interoperability and adapter enumeration/selection, could we maybe point to those samples instead in the documentation? The goal is to upstream the open source samples to [DirectX-Graphics-Samples](https://github.com/microsoft/DirectX-Graphics-Samples) when finished.

You can find the in-progress samples in: [H264 decode sample](https://github.com/sivileri/DirectX-Graphics-Samples/tree/video_sample/Samples/Desktop/D3D12HelloWorld/src/HelloVADecode) similar to `intel/sfcsample` and [VPP Color Conversion + H264 encode](https://github.com/sivileri/DirectX-Graphics-Samples/tree/video_sample/Samples/Desktop/D3D12HelloWorld/src/HelloVAEncode) that show `encode` and `color conversion`

**Pull request merged**
